Kernel/Bus/PCI/DeviceTreeHelpers.cpp

```diff
@@ -80,44 +81,48 @@ ErrorOr<void> configure_devicetree_host_controller(::DeviceTree::Node const& nod
 
     auto const& device_tree = DeviceTree::get();
 
-    if (node.parent() == nullptr)
+    auto const* parent = node.parent();
+
+    if (parent == nullptr)
         return EINVAL;
 
-    auto parent_address_cells = node.parent()->address_cells();
+    auto maybe_ranges = node.ranges();
+    if (!maybe_ranges.is_error()) {
+        auto ranges = maybe_ranges.release_value();
+
+        for (size_t i = 0; i < ranges.entry_count(); i++) {
+            auto range = MUST(ranges.entry(i));
 
-    auto maybe_ranges = node.get_property("ranges"sv);
-    if (maybe_ranges.has_value()) {
-        auto address_cells = node.get_property("#address-cells"sv).value().as<u32>();
-        VERIFY(address_cells == 3); // Additional cell for OpenFirmware PCI address metadata
-        auto size_cells = node.get_property("#size-cells"sv).value().as<u32>();
+            auto raw_pci_address = range.child_bus_address();
+            auto cpu_physical_address = TRY(TRY(parent->translate_child_bus_address_to_root_address(range.parent_bus_address())).as_flatptr());
+            auto range_size = TRY(range.length().as_size_t());
 
-        auto stream = maybe_ranges.value().as_stream();
-        while (!stream.is_eof()) {
-            auto pci_address = TRY(stream.read_value<OpenFirmwareAddress>());
+            if (raw_pci_address.raw().size() != 3 * sizeof(u32))
+                return EINVAL;
 
-            FlatPtr mmio_address = MUST(stream.read_cells(parent_address_cells));
-            u64 mmio_size = MUST(stream.read_cells(size_cells));
+            OpenFirmwareAddress pci_address {};
+            memcpy(&pci_address, raw_pci_address.raw().data(), sizeof(OpenFirmwareAddress));
```

Comment:
Bit unfortunate that this looses a typed read in favour of a memcpy

Reply:
I think it's still nicer than what we had before.
It also handles nested `ranges` now.